File: py/nuclear/find.py
# 查找函数的实现(递归版本)
# def find(data, city_id):
#     cur_nm = ''
#     has_found = False
#
#     def _find(_data, _cid):
#
#         # 闭包需要显式声明变量作用域
#         nonlocal has_found
#         nonlocal cur_nm
#
#         if has_found is True:
#             return cur_nm
#         # 这里是Tuple解构了Tuple[Int, Any]
#         for idx, item in enumerate(_data):
#             # 这里可以看出来，是一个深度优先的遍历
#             # print(item['nm'])
#             if item["id"] == _cid:
#                 has_found = True
#                 cur_nm = item["nm"]
#                 break
#             # 使用item.get, 因为children字段可能不存在，会报错
#             if not has_found and item.get('children') is not None and len(item["children"]):
#                 _find(item["children"], _cid)
#
#     # python没有函数提升?
#     _find(data, city_id)
#
#     return cur_nm

# 查找函数实现, 非递归(实际是用栈实现)
def find(data, city_id):
    # 初始化循环堆栈
    data_stack = []
    # cur_data = data[0]
    # 循环体入栈
    # data_stack.append(cur_data)
    for idx in range(len(data)):
        cur_data = data[idx]
        data_stack.append(cur_data)
        while len(data_stack):
            cur = data_stack[0]
            data_stack = data_stack[1:]
            if cur['id'] == city_id:
                return cur['nm']
            # 这里的遍历顺序，可以看出来是一个横向的, 即广度优先的遍历
            # print(cur['nm'])
            if cur.get('children'):
                for (_, item) in enumerate(cur["children"]):
                    data_stack.append(item)


if __name__ == "__main__":
    source_data = [
        {
            "id": 1,
            "nm": '上海',
            "children": [
                {
                    "id": 11,
                    "nm": '浦东新区',
                    "children": [
                        {
                            "id": 111,
                            "nm": '川沙'
                        }
                    ]
                },
                {
                    "id": 12,
                    "nm": '徐汇区'
                }
            ]
        },
        {
            "id": 2,
            "nm": '广东',
            "children": [
                {
                    "id": 22,
                    "nm": '广州'
                }
            ]
        }
    ]
    print(find(source_data, 111))

File: js/algorithm_nuclear/find.js
// 查找数组中的数据，级联的情况
// 使用递归实现
// 用栈代替递归?
// 使用广度优先的遍历方式实现?

/**
 * 查找一个级联关系的数据
 */

const sourceData = [
    {
        id: 1,
        nm: '上海',
        children: [
            {
                id: 11,
                nm: '浦东新区',
                children: [
                    {
                        id: 111,
                        nm: '川沙'
                    }
                ]
            },
            {
                id: 12,
                nm: '徐汇区'
            }
        ]
    },
    {
        id: 2,
        nm: '广东',
        children: [
            {
                id: 22,
                nm: '广州'
            }
        ]
    }
]

/**
 * 用栈实现递归算法的非递归版本
 * 递归实现存在一个中断
 * 栈的实现也存在一个中断, 那就是栈为空,
 * 在这个栈为空的前提下，以while和push来模拟递归调用, 以pop来减少堆栈数据
 */
function find(data, id) {
    let stack = [];
    for (let i = 0; i < data.length; i++) {
        let curDeep = data[i];
        if (curDeep.id == id) {
            return curDeep.nm
        }
        if (curDeep.children) {
            stack.push(...curDeep.children);
            while(stack.length) {
                let curPop = stack.pop();
                if (curPop.id == id) {
                    return curPop.nm
                }
                if (curPop.children) {
                    stack.push(...curPop.children)
                }
            }
        }
    }
}

/**
 * find(data, 111) // 川沙
 * find(data, 11) // 浦东新区
 */

/**
 * 递归实现, 这个写法没有hasFound 变量中断的话，总是会把循环体走完
 */
// function find(data, id) {

//     let curNm = '';
//     let hasFound = false;

//     _find(data, id)

//     function _find(fData, fid) {
//         for (let i = 0; i < fData.length; i++) {
//             // 可以添加变量标志的方式来限制查询次数
//             if (hasFound) {
//                 break
//             }
//             // console.log(fData[i].nm)
//             if (fData[i].id == fid) {
//                 curNm = fData[i].nm
//                 hasFound = true;
//             }
//             if (fData[i].children && fData[i].children.length) {
//                 _find(fData[i].children, id);
//             }
//         }
//     }
//     return curNm
// }

const nm = find(sourceData, 11);

console.log(nm);